File: src/python/place_registry.py
#!/usr/bin/env python3
"""
Python access to the canonical place registry (data/places.json).

The mirror of src/js/maps/place-registry.js — same normalisation (norm_place)
and the SAME find_places_in_text algorithm (longest-first, skip names contained
in an already-matched longer name), so the categorizer, the validator and the
browser all resolve a town to the same canonical place.
"""
import json
import os

from categorize import norm_place

_BASE = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
_PLACES_PATH = os.path.join(_BASE, 'data', 'places.json')

_by_norm = None
_sorted_keys = None


def load_places(path=None):
    """Load and cache {norm_key: entry}. entry = {display,lat,lng,confidence,...}."""
    global _by_norm, _sorted_keys
    if _by_norm is not None and path is None:
        return _by_norm
    p = path or _PLACES_PATH
    with open(p, encoding='utf-8') as f:
        doc = json.load(f)
    by_norm = doc.get('places', {})
    if path is None:
        _by_norm = by_norm
        _sorted_keys = sorted(by_norm.keys(), key=len, reverse=True)
    return by_norm


def _keys_longest_first():
    global _sorted_keys
    if _sorted_keys is None:
        load_places()
    return _sorted_keys


def lookup(name):
    """Exact registry entry for a name (via norm_place), or None."""
    if not name:
        return None
    return load_places().get(norm_place(name))


def coord_of(name):
    e = lookup(name)
    if e and e.get('lat') is not None:
        return (e['lat'], e['lng'])
    return None


# Arabic single-letter proclitics that may attach before a town name
# (و=and, ف=so/then, ب=in/with, ك=like, ل=for). A name may be preceded by ONE of
# these at a word boundary (e.g. "والعديسة"); anything else must be a boundary.
_PROCLITICS = set('وفبكل')


def _is_ar_letter(c):
    return 'ء' <= c <= 'ي'


def _word_match(nt, k):
    """True if `k` occurs in `nt` as a whole word — bounded by non-letters, with
    an optional single proclitic before it. Prevents short names ('دان') from
    matching inside longer words ('بلدان', 'ميدان')."""
    start = 0
    n = len(nt)
    while True:
        i = nt.find(k, start)
        if i < 0:
            return False
        end = i + len(k)
        left_ok = (i == 0) or (not _is_ar_letter(nt[i - 1])) or (
            nt[i - 1] in _PROCLITICS and (i - 1 == 0 or not _is_ar_letter(nt[i - 2])))
        right_ok = (end >= n) or (not _is_ar_letter(nt[end]))
        if left_ok and right_ok:
            return True
        start = i + 1


def find_places_in_text(text):
    """Every canonical place whose normalised name occurs in `text` as a WHOLE
    WORD (proclitics allowed). Longest-first; a name fully contained in an
    already-matched longer name is skipped (so 'جبيل' inside 'بنت جبيل' is not
    double-counted). Returns canonical norm keys, in match order. MULTI-TARGET aware.
    """
    if not text:
        return []
    nt = norm_place(text)
    hits = []
    for k in _keys_longest_first():
        if not k or k not in nt:
            continue
        if any(k in h for h in hits):
            continue
        if _word_match(nt, k):
            hits.append(k)
    return hits

File: src/js/maps/place-registry.js
/* === src/js/maps/place-registry.js === */
/*
 * The SINGLE source of truth for town -> coordinate lookups in the browser.
 * Loads data/places.json (built by src/python/build_places.py from all the old
 * divergent tables) and resolves names through normalizePlace, so the map, the
 * dashboards and search all place/count a town identically — and identically to
 * the Python pipeline (categorize.py / validate_consistency.py).
 *
 * Replaces the per-view hardcoded tables: bayanat-map opCoords, bayanat-dash
 * locNames, siren-auto-map cityCoords, sirens-dash regionKw.
 */

import { normalizePlace } from '../util/normalize-ar.js';

var _placesPromise = null;   // raw {norm_key: entry}
var _byNorm = null;          // resolved map once loaded
var _sortedKeys = null;      // norm keys, longest-first (for findInText)

export function loadPlaces() {
  if (_placesPromise) return _placesPromise;
  var candidates = ['data/places.json', '../data/places.json', './data/places.json'];
  _placesPromise = (function tryNext(i) {
    if (i >= candidates.length) return Promise.resolve({});
    return fetch(candidates[i])
      .then(function (r) { return r.ok ? r.json() : tryNext(i + 1); })
      .catch(function () { return tryNext(i + 1); });
  })(0).then(function (doc) {
    _byNorm = (doc && doc.places) ? doc.places : {};
    _sortedKeys = Object.keys(_byNorm).sort(function (a, b) { return b.length - a.length; });
    return _byNorm;
  });
  return _placesPromise;
}

// Exact lookup of a single name (already loaded). Returns the registry entry
// {display, lat, lng, confidence, hasPolygon, aliases} or null.
export function lookupPlace(name) {
  if (!_byNorm || !name) return null;
  return _byNorm[normalizePlace(name)] || null;
}

export function coordOf(name) {
  var e = lookupPlace(name);
  return (e && e.lat != null) ? [e.lat, e.lng] : null;
}

// Arabic single-letter proclitics (و ف ب ك ل) that may attach before a name.
var _PROCLITICS = 'وفبكل';
function _isArLetter(c) { return c >= 'ء' && c <= 'ي'; }

// Whole-word match (mirror of place_registry.py _word_match): `k` bounded by
// non-letters, with an optional single proclitic before it — so short names
// ('دان') don't match inside longer words ('بلدان', 'ميدان').
function _wordMatch(nt, k) {
  var start = 0, n = nt.length;
  for (;;) {
    var i = nt.indexOf(k, start);
    if (i < 0) return false;
    var end = i + k.length;
    var leftOk = (i === 0) || !_isArLetter(nt.charAt(i - 1)) ||
      (_PROCLITICS.indexOf(nt.charAt(i - 1)) !== -1 && (i - 1 === 0 || !_isArLetter(nt.charAt(i - 2))));
    var rightOk = (end >= n) || !_isArLetter(nt.charAt(end));
    if (leftOk && rightOk) return true;
    start = i + 1;
  }
}

// MULTI-TARGET aware: every canonical place whose normalized name occurs in the
// normalized text as a WHOLE WORD (proclitics allowed). Longest-first; a name
// fully contained in a longer matched name is skipped (won't count "جبيل" inside
// "بنت جبيل"). Returns canonical norm keys. Mirror of Python find_places_in_text.
export function findPlacesInText(text) {
  if (!_byNorm || !text) return [];
  var nt = normalizePlace(text);
  var hits = [];
  for (var i = 0; i < _sortedKeys.length; i++) {
    var k = _sortedKeys[i];
    if (!k || nt.indexOf(k) === -1) continue;
    var contained = false;
    for (var j = 0; j < hits.length; j++) {
      if (hits[j].indexOf(k) !== -1) { contained = true; break; }
    }
    if (!contained && _wordMatch(nt, k)) hits.push(k);
  }
  return hits;
}

export function placeEntries() { return _byNorm || {}; }

// Test seam: inject the registry without fetch (used by the cross-language
// agreement check in validate_consistency.py running under node).
export function __setRegistryForTest(places) {
  _byNorm = places || {};
  _sortedKeys = Object.keys(_byNorm).sort(function (a, b) { return b.length - a.length; });
}
